pkg/dmaasbackup/schedule.go

```diff
@@ -0,0 +1,175 @@
+/*
+Copyright 2020 The MayaData Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    https://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dmaasbackup
+
+import (
+	"time"
+
+	velerov1api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	velerobuilder "github.com/vmware-tanzu/velero/pkg/builder"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/mayadata-io/dmaas-operator/pkg/apis/mayadata.io/v1alpha1"
+	"github.com/pkg/errors"
+	"github.com/robfig/cron"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+)
+
+const (
+	// dmaasScheduleSuffix suffix for schedule name generated by dmaas operator
+	dmaasScheduleSuffix = "dmaas"
+)
+
+func (d *dmaasBackup) processSchedule(obj *v1alpha1.DMaaSBackup) error {
+	// check if crontime is mentioned
+	if obj.Spec.PeriodicFullBackupCfg.CronTime == "" {
+		return d.processNonFullBackupSchedule(obj)
+	}
+
+	d.logger.Debug("Processing fullbackup")
+
+	cr, err := cron.ParseStandard(obj.Spec.PeriodicFullBackupCfg.CronTime)
+	if err != nil {
+		return errors.Wrapf(err, "failed to parse cronTime")
+	}
+
+	activeSchedule := getActiveVeleroSchedule(obj)
+
+	isDue, nextDue := getNextDue(cr, activeSchedule, d.clock.Now())
+	if isDue {
+		d.logger.Debug("Creating full backup")
+
+		// before creating a new schedule, we will delete the old schedule
+		// upon successful deletion we will create a new schedule
+		if activeSchedule != nil {
+			err = d.scheduleClient.Delete(
+				activeSchedule.ScheduleName,
+				&metav1.DeleteOptions{},
+			)
+			if err != nil {
+				if !apierrors.IsNotFound(err) {
+					d.logger.WithError(err).
+						Errorf("failed to delete schedule=%s", activeSchedule.ScheduleName)
+					return errors.Wrapf(err,
+						"failed to delete schedule=%s", activeSchedule.ScheduleName)
+				}
+			}
+			activeSchedule.Status = v1alpha1.Deleted
+			d.logger.Infof("Schedule=%s deleted", activeSchedule.ScheduleName)
+		}
+
+		// create a new schedule
+		newSchedule, err := d.createSchedule(obj)
```

Comment:
same comment applies as mentioned in processNonFullBackupSchedule due to duplicate in behavior